datafusion/common/src/tree_node.rs

```diff
@@ -174,6 +182,70 @@ pub trait TreeNode: Sized {
         })
     }
 
+    /// Implements the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) for
```

Comment:
Here is the new TreeNode API to rewrite the nodes in place

I also tried  `Transformed<&mut Node>`, but this got caught up in the borrow checker. Returning `Transformed<()>` seemed to work and made using this API feasible

@peter-toth  and @berkaysynnada  I would love any feedback on this API that you may have